rmw_connextdds_common/src/common/rmw_impl_waitset_dds.cpp

```diff
@@ -721,50 +702,35 @@ RMW_Connext_WaitSet::invalidate(RMW_Connext_Condition * const condition)
 {
   std::unique_lock<std::mutex> lock(this->mutex_internal);
 
+  // Ideally, we would consider an error if the WaitSet is not in FREE or
+  // ACQUIRING state, because it signals parallel wait() and delete(). This
+  // should probably be considered an "application error", but it seems like
+  // some packages might not be making the same assumption, so for now, we wait
+  // for the waitset to be free. This might block a thread indefinitely
+  // (if the parallel wait has infinite timeout and never returns).
+  while (this->state != RMW_CONNEXT_WAITSET_FREE) {
+    RMW_CONNEXT_LOG_DEBUG_A(
+      "waiting for waitset to become available for invalidation: "
+      "waitset=%p, condition=%p",
+      static_cast<void *>(this), static_cast<void *>(condition))
+    this->state_cond.wait(lock);
+  }
```

Comment:
The problem here is an object deletion (e.g. a subscription, or a publisher, or a guard condition) in parallel with a `wait()` on a WaitSet that was previously used to wait on the same object. This caused the object to be cached in the WaitSet instance and this caused is trying to make sure that the WaitSet doesn't try to access the object again after it's deleted (and this function returns).

Btw, this is the "concurrency nightmare" I mentioned in my initial post.

The original implementation was stricter and threw an error if it detected that a WaitSet had entered a `wait()` with a deleted object still attached to it. I thought this was only "extremely defensive coding" on my part, because I had your same assumption on the validity of such use case (not supported), but it seems it was being triggered by `nav2` tests (see #21, coming from `gzserver`).

I tried running with this code, but I think `gzserver` was ending up in a crash, hence why I stopped trying to fix it, and just switched to the alternative WaitSet implementation.

Reply:
If we've switched to another implementation, why commit this (partial changes)?

Reply:
The reason for the partial changes in this PR is that when I started investigating #21, I first tried to fix the DDS-based implementation, before realizing that it was causing too much CPU usage anyway, and that fixing the C++ std one would have probably ended up getting us a "better return for the investment".

I could have just deleted the DDS-based implementation, but at this point I've switched between the two implementations back and forth several times, and I didn't want to delete it again, just to have to exhume sometime later in the future.

At the same time though, I doubt we will be reconsidering the implementation in the near future, at least not until either:
- I find time to investigate the performance bottlenecks in more detail, and possibly identify improvements in the Connext libraries.
- ROS 2 introduces `attach()/detach()` and we can re-evaluate the code without all the "hackish" optimizations to try to cope with the lack of these functions in the API.

So maybe, the best thing would be to preserve this implementation in a (dead for now) branch, and clean up the code to only include the one that is actually used.

I can modify the PR to this end, and create a personal branch to preserve the DDS-based WaitSet implementation.